multisense_ros/include/multisense_ros/point_cloud_utilities.h

```diff
@@ -34,44 +34,123 @@
+
+    point_cloud.point_step += sizeof(ColorT);
+}
+
+template <typename ColorT>
+typename std::enable_if<std::is_same<ColorT, void>::value, void>::type
+initPointcloudColorChannel(sensor_msgs::PointCloud2& point_cloud,
+                           const size_t offset,
+                           const std::string& color_channel)
+{
+    (void) point_cloud;
+    (void) offset;
+    (void) color_channel;
+    return;
+}
+
+template <typename PointT, typename ColorT>
+void initializePointcloud (sensor_msgs::PointCloud2& point_cloud,
+                           const ros::Time& stamp,
+                           const size_t width,
+                           const size_t height,
+                           const bool dense,
+                           const std::string& frame_id,
+                           const std::string& color_channel)
+{
+    const auto point_datatype = messageFormat<PointT>();
+
+    const bool has_color = not std::is_same<ColorT, void>::value;
 
-    sensor_msgs::PointCloud2 point_cloud;
     point_cloud.is_bigendian    = (htonl(1) == 1);
     point_cloud.is_dense        = dense;
-    point_cloud.point_step      = 4 * sizeof(T);
+    point_cloud.point_step = 0;
     point_cloud.header.frame_id = frame_id;
-    point_cloud.fields.resize(4);
+    point_cloud.header.stamp = stamp;
+    point_cloud.width = width;
+    point_cloud.height = height;
+
+    if (has_color)
+    {
+        point_cloud.fields.resize(4);
+    }
+    else
+    {
+        point_cloud.fields.resize(3);
+    }
+
     point_cloud.fields[0].name     = "x";
-    point_cloud.fields[0].offset   = 0;
+    point_cloud.fields[0].offset   = point_cloud.point_step;
     point_cloud.fields[0].count    = 1;
-    point_cloud.fields[0].datatype = datatype;
+    point_cloud.fields[0].datatype = point_datatype;
+    point_cloud.point_step += sizeof(PointT);
+
     point_cloud.fields[1].name     = "y";
-    point_cloud.fields[1].offset   = sizeof(T);
+    point_cloud.fields[1].offset   = point_cloud.point_step;
     point_cloud.fields[1].count    = 1;
-    point_cloud.fields[1].datatype = datatype;
+    point_cloud.fields[1].datatype = point_datatype;
+    point_cloud.point_step += sizeof(PointT);
+
     point_cloud.fields[2].name     = "z";
-    point_cloud.fields[2].offset   = 2 * sizeof(T);
+    point_cloud.fields[2].offset   = point_cloud.point_step;
     point_cloud.fields[2].count    = 1;
-    point_cloud.fields[2].datatype = datatype;
-    point_cloud.fields[3].name     = color_channel;
-    point_cloud.fields[3].offset   = 3 * sizeof(T);
-    point_cloud.fields[3].count    = 1;
-    point_cloud.fields[3].datatype = datatype;
+    point_cloud.fields[2].datatype = point_datatype;
+    point_cloud.point_step += sizeof(PointT);
+
+    if (has_color)
+    {
+        initPointcloudColorChannel<ColorT>(point_cloud,
+                                           point_cloud.point_step,
+                                           color_channel);
+    }
```

Comment:
If there was a way to make this a compile-time conditional, that would save the templated functions above in the cae where ColorT is void

Reply:
Do you create another parallel function for color pointclouds? You could also use things like std::enable_if but that quickly becomes hard to manage.